### PR TITLE
Conv2d backward

### PR DESCRIPTION
**Pull request by koenvanderveen**
This PR implements computing the gradients for the convolutional layer. Since I had a some trouble with reshaping/transposing I also ended up implementing a unit test for im2col.

### CODE REVIEW

**Comment by mortendahl** [comment]
@koenvanderveen want to get the linting errors fixed (`make lint`) so we can get this merged up?

**Comment by justin1121** [comment]
Hey @koenvanderveen. Awesome PR! The tests were broken for an unrelated thing to your stuff. If you want to merge with master and fix conflicts and make sure the tests will pass I'm sure we can get this merged up.

File: tensorflow_encrypted/protocol/pond.py [hunk 1]
```diff
@@ -794,6 +817,34 @@ def conv2d(self, x, w, strides, padding):
 
         return z
 
+    def conv2d_bw(self, x, d_y, w_shape, strides, padding):
+        node_key = ('conv2d', x, d_y, w_shape)
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/pond.py]: copy/paste error in `conv2d` and not `conv2d_bw`?

File: tensorflow_encrypted/layers/convolution.py [hunk 2]
```diff
@@ -80,24 +80,25 @@ def backward(self, d_y, learning_rate):
...
 
-        if self.model.layers.index(self) != 0:
-            W_reshaped = self.weights.reshape(n_filter, -1).transpose()
-            dout_reshaped = d_y.transpose(1, 2, 3, 0).reshape(n_filter, -1)
-            dx = W_reshaped.dot(dout_reshaped).col2im(imshape=self.cached_input_shape,
-                                                      field_height=h_filter,
-                                                      field_width=w_filter,
-                                                      padding=self.padding,
-                                                      stride=self.strides)
+        dx = None
```
Comment (mortendahl) [inline tensorflow_encrypted/layers/convolution.py]: is this on purpose, including the commented code below?

File: tensorflow_encrypted/protocol/pond.py [hunk 3]
```diff
@@ -1454,6 +1505,64 @@ def _reveal_masked(prot, x):
...
+
+        with tf.device(prot.server_0.device_name):
+            a0_col = a0.im2col(h_filter, w_filter, padding, strides)
+            alpha_on_0_col = alpha_on_0.im2col(h_filter, w_filter, padding, strides)
+
+        with tf.device(prot.server_1.device_name):
+            a1_col = a1.im2col(h_filter, w_filter, padding, strides)
+            alpha_on_1_col = alpha_on_1.im2col(h_filter, w_filter, padding, strides)
+
+    x_unmasked_col = prot.im2col(x_masked.unmasked, h_filter, w_filter, padding, strides)
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/pond.py]: missing one indent here to get it under the `im2col` name scope (makes more sense in TensorBoard when they're nested)

File: tensorflow_encrypted/layers/convolution.py (hunk 2 above)
Comment (koenvanderveen) [reply tensorflow_encrypted/layers/convolution.py]: yes it is on purpose, it was pseudocode for when I start with implementing the dx computation from the privateml repo (probably tomorrow). But it does not need to be in master.. Do you want me to remove it?

File: tensorflow_encrypted/protocol/pond.py (hunk 3 above)
Comment (koenvanderveen) [reply tensorflow_encrypted/protocol/pond.py]: only this line correct? not the next one

File: tensorflow_encrypted/layers/convolution.py (hunk 2 above)
Comment (mortendahl) [reply tensorflow_encrypted/layers/convolution.py]: no worries, just wanted to know! you'll keep working on this branch?

Comment (koenvanderveen) [reply tensorflow_encrypted/layers/convolution.py]: yes

**Comment by gavinuhma** [comment]
+1
